docs/architecture/adr/ADR044-dev-namespaces.md

```diff
@@ -0,0 +1,74 @@
+# ADR044: Dev namespaces
+
+## Status
+
+Pending
+
+## Context
+
+Following the retirement of gsp-local the service teams have no way of testing
+changes before a merge to master. So the feedback cycle is very slow, and
+potentially risky. We need to enable devs to get more rapid feedback through use
+of namespaces that allow for deployments from arbitrary branches or from local
+`kubectl apply` actions.
+
+This presents several security problems that need to be addressed.
+
+The in-cluster concourse:
+
+* must not be able to create or edit daemonsets
+* must not be able to create pods that use host networking, or run in privildged
+  mode
+* must not be authorised to create any "cluster" scoped resources (e.g.
+  clusterrole, custom resources etc.)
+
+For harbor:
+
+* a dev in one namespace must not be authorised to push, edit or delete images
+  relating to namespaces outside their own
+* notary signing keys must be namespace-scoped
+
+For external-dns:
+
+* a namespace must not be able to hijack the DNS entries of another namespace
+
+For istio:
+
+* any istio resources deployed as part of a namespace must not have any impact
+  on other namespaces
+
+For CloudHSM access:
+
+* connectivity only enabled via `-cluster-config`, as with "prod" namespaces and
+  pods
+* credentials should be different for each namespace, and not the same as those
+  used for the "production" instances of the application
```

Comment:
Another things I can think of:

----

For Pods of any kind:

* `hostPID` must be `false`, to prevent pods accessing arbitrary processes on the host
* `hostNetwork` must be `false`, to prevent pods evading network security policies and accessing privileged network endpoints
* probably lots of other things I haven't thought of to prevent pods doing privileged things

(is the answer to this [PodSecurityPolicies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/)?)

Reply:
I have included this (and related) in another ADR (and renumbered this one for consistency reasons).